plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java

```diff
@@ -1218,13 +1246,31 @@ public void addColumn(ConnectorSession session, ConnectorTableHandle tableHandle
         try {
             long commitVersion = handle.getReadVersion() + 1;
 
+            ImmutableMap.Builder<String, Map<String, Object>> columnMetadata = ImmutableMap.builder();
+            columnMetadata.putAll(getColumnsMetadata(handle.getMetadataEntry()));
+            Map<String, String> configuration = new HashMap<>(handle.getMetadataEntry().getConfiguration());
+            AtomicInteger maxColumnId;
+            boolean assignIdAndPhysicalName;
+            switch (columnMappingMode) {
+                case NONE -> {
+                    maxColumnId = new AtomicInteger();
+                    assignIdAndPhysicalName = false;
+                }
+                case ID, NAME -> {
+                    maxColumnId = new AtomicInteger(getMaxColumnId(handle.getMetadataEntry()));
+                    assignIdAndPhysicalName = true;
+                }
+                default -> throw new IllegalArgumentException("Unexpected column mapping mode: " + columnMappingMode);
+            }
+            columnMetadata.put(newColumnMetadata.getName(), ImmutableMap.of());
```

Comment:
Let's fill the `columnMetadata` here and not in `DeltaLakeSchemaSupport` .
Doing this would spare us from passing `columnMappingMode`, `maxColumnId`, `newColumnName` to `io.trino.plugin.deltalake.transactionlog.DeltaLakeSchemaSupport#serializeSchemaAsJson`

Reply:
I think it could be doable with a special method in `DeltaLakeSchemaSupport`.

Passing around `maxColumnId` makes it hard to follow when it actually gets modified.

The call to `io.trino.plugin.deltalake.transactionlog.DeltaLakeSchemaSupport#serializeColumnType` increments `maxColumnId` and also `serializeSchemaAsJson` increments it. The order of the calls is also important.

I'd advocate here that the method `serializeSchemaAsJson` does not change the state of `maxColumnId` and it does only what is supposed to do - json serialization and no extra mutation logic.